dev/src/bulk-writer.ts

```diff
@@ -620,16 +744,33 @@ export class BulkWriter {
    */
   async flush(): Promise<void> {
     this.verifyNotClosed();
-    const trackedBatches = this.batchQueue;
-    const writePromises = trackedBatches.map(batch => batch.awaitBulkCommit());
-    this.sendReadyBatches();
-    await Promise.all(writePromises);
+    await this._flush();
+  }
+
+  private async _flush(): Promise<void> {
+    const trackedBatches = this._batchQueue;
+    trackedBatches.map(batch => batch.markReadyToSend());
+    let retriesCompleted = false;
+    while (!retriesCompleted) {
+      const before = this._pendingBatches.size;
+      this.sendReadyBatches();
+      const after = this._pendingBatches.size;
```

Comment:
If a user adds a new batch, this will send the new user writes as well, won't it? I think this creates the "flush loops over all writes while any write is pending and is impossible to understand API" that I had in one of my earlier revisions. I think what you want is a single commit of all non-retry batches and then a final loop that drains the retry batches.